State: Adds actions and tests for Jetpack SSO

ebinnion opened the pull request:
This PR will be used to add actions to global state that will be used for Jetpack SSO on the `/jetpack/sso` route. For testing purposes, I updated the actions to use promises, so I also updated the reducer and any related tests.

To test:
- Checkout `add/jetpack-sso-actions` branch
- `npm run test-client -- --grep "state jetpack-connect"`

cc @lezama and @gwwar for review.

gwwar commented on client/state/jetpack-connect/actions.js:
```diff
@@ -200,5 +205,53 @@ export default {
+
+			return wpcom.undocumented().jetpackValidateSSONonce( siteId, ssoNonce ).then( ( data ) => {
+				dispatch( {
+					type: JETPACK_CONNECT_SSO_VALIDATION_RECEIVE,
```
This is optional, but a common pattern is to have 4 action types for a given request.
```
JETPACK_CONNECT_SSO_VALIDATION_REQUEST,
JETPACK_CONNECT_SSO_VALIDATION_RECEIVE,
JETPACK_CONNECT_SSO_VALIDATION_SUCCESS,
JETPACK_CONNECT_SSO_VALIDATION_ERROR
```

`JETPACK_CONNECT_SSO_VALIDATION_REQUEST` is dispatched like you have here immediately. While `JETPACK_CONNECT_SSO_VALIDATION_RECEIVE` and `JETPACK_CONNECT_SSO_VALIDATION_SUCCESS` are dispatched on success. Normally the receive action has the data payload, while we can use the `JETPACK_CONNECT_SSO_VALIDATION_SUCCESS` action to keep track of whether our action has completed. On error, we dispatch the `JETPACK_CONNECT_SSO_VALIDATION_ERROR` action with the error payload.

This may seem like an excessive number of action types, but it helps keep the reducer data simple. (See https://github.com/Automattic/wp-calypso/blob/master/client/state/sites/media-storage/reducer.js for an example) In this case, since we're only interested in a boolean value, I'd probably suggest using:
```
JETPACK_CONNECT_SSO_VALIDATION_REQUEST,
JETPACK_CONNECT_SSO_VALIDATION_SUCCESS,
JETPACK_CONNECT_SSO_VALIDATION_ERROR
```

gwwar commented on client/state/jetpack-connect/reducer.js:
```diff
@@ -123,19 +123,11 @@ export function jetpackSSO( state = {}, action ) {
-		case SERIALIZE:
+			return Object. assign( state, { isAuthorizing: false, authorizationError: action.error, ssoUrl: action.ssoUrl } );
 		case SERIALIZE:
+		case DESERIALIZE:
```
👍

gwwar replied:
While we're in this file, any chance we could fix https://github.com/Automattic/wp-calypso/blob/b6a8a750d826c3e2c9b47ac26e2e46992aef595c/client/state/jetpack-connect/reducer.js#L47 and https://github.com/Automattic/wp-calypso/blob/b6a8a750d826c3e2c9b47ac26e2e46992aef595c/client/state/jetpack-connect/reducer.js#L112

Steps would be
1. To double check if we want to persist these sub-trees
2. If it makes sense to persist, add a schema to guard against data drift
3. If it doesn't make sense, we'll want to return default state like we do here.

Also see: https://github.com/Automattic/wp-calypso/blob/master/docs/our-approach-to-data.md#data-persistence--2754-

If you have questions, feel free to ping me.

gwwar commented:
👍 The new action tests look good here.

ebinnion commented:
Thank you for the review @gwwar! I addressed your comments about using more standardized action names.

In regards to the second comment, I'd rather not address it here. I created #5330 for adding schema that way we can loop in @johnHackworth and team luna.

lezama commented:
This LGTM @ebinnion!

:shipit:

lezama commented on client/lib/wpcom-undocumented/lib/undocumented.js:
```diff
@@ -186,6 +186,20 @@ Undocumented.prototype.jetpackAuthorize = function( siteId, code, state, redirec
+};
+
+Undocumented.prototype.jetpackAuthorizeSSONonce = function( siteId, ssoNonce, fn ) {
+	debug( '/jetpack-blogs/:site_id:/sso-validate query' );
```
this should say `'/jetpack-blogs/:site_id:/sso-validate query'`

ebinnion replied:
Good eyes! I'll get that fixed up and deployed. 😄
